awx/main/scheduler/task_manager.py

```diff
@@ -523,6 +523,7 @@ def timeout_approval_node(self):
         workflow_approval = WorkflowApproval.objects.filter(status='pending').prefetch_related('workflow_approval_template')
         now = tz_now()
         for task in workflow_approval:
+            # TODO: copy the timeout to the job itself at launch time, not the template
```

Comment:
@beeankha,

We should *copy* the timeout field _from_ the `WorkflowApprovalTemplate` _to_ the `WorkflowApproval` that gets created.  That way if the template is deleted, or the timeout is _edited_ on the template, it won't affect `pending` approvals.

We already have a pattern for this sort of things in our `UnifiedJobTemplate` -> `UnifiedJob` paradigm.

This is the method that *creates* a unified job:

https://github.com/ansible/awx/blob/devel/awx/main/models/unified_jobs.py#L331

One of the goals of this method is to identify *which fields* from the UJT should be copied to the UJ:

https://github.com/ansible/awx/blob/devel/awx/main/models/unified_jobs.py#L344

Child classes define which fields should be copied, and you'll want to do the same for the new `WorkflowApproval` code:

1.  Add a new `timeout` field/column to `WorkflowApproval` that is identical to the one on `WorkflowApprovalTemplate`.

2.  Define a `_get_unified_job_field_names` method on `WorkflowApproval` that _includes_ the new `timeout` field.  Here are some examples of other UJ implementations that do this:

https://github.com/ansible/awx/blob/devel/awx/main/models/jobs.py#L260
https://github.com/ansible/awx/blob/devel/awx/main/models/jobs.py#L1122

3.  Update _this_ code to reference the timeout field from the `WorkflowApproval` itself, *not* the template.
4.  Anything else that might be necessary (exercise left to the reader).

Reply:
> Add a new timeout field/column to WorkflowApproval that is identical to the one on WorkflowApprovalTemplate.

Alternative pattern that we've been rolling with for JT-job relationship:

Introduce an `ApprovalOptions` abstract class, and have both the approval JTs and approval jobs inherit from this. When writing the `_get_unified_job_field_names`, you can start out including all field names from the "options" class.

Either way is fine. I only suggest this because the pattern already exists, so helps to make it easy to follow / emulate.

Reply:
@AlanCoding I thought about this, but also thought it might be a bit much since we've only got one field at this point (and I don't anticipate any others in the near future).